Token leaking into logfile

I noticed tokens still leaking into logfiles in the current HEAD of the devel branch.

Here are current examples (where I replace my token's signature with `READACTED` manually):

```
Dumping request: PUT /first/namespace/token_gen/test1171569942 HTTP/1.1\r\nHost: F4HP7QL65F.local:61632\r\nUser-Agent: pelican-client/dev\r\nTransfer-Encoding: chunked\r\nAuthorization: Bearer eyJhbGciOiJFUzI1NiIsImtpZCI6IjBabFNVdi10ZXJhc0RFZmY5a2U2SGtHRU1JRFByR3FqWS1XVEpFbmJMcG8iLCJ0eXAiOiJKV1QifQ.eyJhdWQiOlsiaHR0cHM6Ly93bGNnLmNlcm4uY2gvand0L3YxL2FueSJdLCJleHAiOjE3MjY3NTUwNjcsImlhdCI6MTcyNjc1MTQ2NywiaXNzIjoiaHR0cHM6Ly9GNEhQN1FMNjVGLmxvY2FsOjYxNjMxIiwianRpIjoiRGFZVk5sZlVaSmNTMUNqVW1ZSmpPdyIsIm5iZiI6MTcyNjc1MTQ2Nywic2NvcGUiOiJzdG9yYWdlLmNyZWF0ZTovdG9rZW5fZ2VuL3Rlc3QxMTcxNTY5OTQyIiwic3ViIjoiY2xpZW50X3Rva2VuIiwid2xjZy52ZXIiOiIxLjAifQ.REDACTED\r\nAccept-Encoding: gzip\r\n\r\n
240919 08:11:07 20995 unknown.3:33@[::1] Pss_Stat: url=pelican://p0@F4HP7QL65F.local:61631//first/namespace/token_gen/test1171569942?&authz=Bearer%20eyJhbGciOiJFUzI1NiIsImtpZCI6IjBabFNVdi10ZXJhc0RFZmY5a2U2SGtHRU1JRFByR3FqWS1XVEpFbmJMcG8iLCJ0eXAiOiJKV1QifQ.eyJhdWQiOlsiaHR0cHM6Ly93bGNnLmNlcm4uY2gvand0L3YxL2FueSJdLCJleHAiOjE3MjY3NTUwNjcsImlhdCI6MTcyNjc1MTQ2NywiaXNzIjoiaHR0cHM6Ly9GNEhQN1FMNjVGLmxvY2FsOjYxNjMxIiwianRpIjoib25FOUlQS0FnM0V2Z0Q0YkQ2MkZyUSIsIm5iZiI6MTcyNjc1MTQ2Nywic2NvcGUiOiJzdG9yYWdlLnJlYWQ6L3Rva2VuX2dlbi90ZXN0MTE3MTU2OTk0MiIsInN1YiI6ImNsaWVudF90b2tlbiIsIndsY2cudmVyIjoiMS4wIn0.REDACTED&pelican.timeout=9.5s
[2024-09-19 08:11:07.679632 -0500][Debug  ][ExDbgMsg          ] [p0@F4HP7QL65F.local:61631] MsgHandler created: 0x26e0e3d0 (message: kXR_stat (path: /first/namespace/token_gen/test1171569942?authz=REDACTED%20eyJhbGciOiJFUzI1NiIsImtpZCI6IjBabFNVdi10ZXJhc0RFZmY5a2U2SGtHRU1JRFByR3FqWS1XVEpFbmJMcG8iLCJ0eXAiOiJKV1QifQ.eyJhdWQiOlsiaHR0cHM6Ly93bGNnLmNlcm4uY2gvand0L3YxL2FueSJdLCJleHAiOjE3MjY3NTUwNjcsImlhdCI6MTcyNjc1MTQ2NywiaXNzIjoiaHR0cHM6Ly9GNEhQN1FMNjVGLmxvY2FsOjYxNjMxIiwianRpIjoib25FOUlQS0FnM0V2Z0Q0YkQ2MkZyUSIsIm5iZiI6MTcyNjc1MTQ2Nywic2NvcGUiOiJzdG9yYWdlLnJlYWQ6L3Rva2VuX2dlbi90ZXN0MTE3MTU2OTk0MiIsInN1YiI6ImNsaWVudF90b2tlbiIsIndsY2cudmVyIjoiMS4wIn0.REDACTED&pelican.timeout=9.5s, flags: none) ).
240919 08:11:07 20995 Posix_Open: [ERROR] Operation not supported open pelican://u33@F4HP7QL65F.local:61631//first/namespace/token_gen/test1171569942?&authz=Bearer%20eyJhbGciOiJFUzI1NiIsImtpZCI6IjBabFNVdi10ZXJhc0RFZmY5a2U2SGtHRU1JRFByR3FqWS1XVEpFbmJMcG8iLCJ0eXAiOiJKV1QifQ.eyJhdWQiOlsiaHR0cHM6Ly93bGNnLmNlcm4uY2gvand0L3YxL2FueSJdLCJleHAiOjE3MjY3NTUwNjcsImlhdCI6MTcyNjc1MTQ2NywiaXNzIjoiaHR0cHM6Ly9GNEhQN1FMNjVGLmxvY2FsOjYxNjMxIiwianRpIjoib25FOUlQS0FnM0V2Z0Q0YkQ2MkZyUSIsIm5iZiI6MTcyNjc1MTQ2Nywic2NvcGUiOiJzdG9yYWdlLnJlYWQ6L3Rva2VuX2dlbi90ZXN0MTE3MTU2OTk0MiIsInN1YiI6ImNsaWVudF90b2tlbiIsIndsY2cudmVyIjoiMS4wIn0.REDACTED&pelican.timeout=9.5s
```

One of these looks like a bad regex in XrdCl.  The others look legit.

--- a/src/XrdHttp/XrdHttpReq.cc
+++ b/src/XrdHttp/XrdHttpReq.cc
@@ -611,7 +611,7 @@ bool XrdHttpReq::Redir(XrdXrootd::Bridge::Context &info, //!< the result context
     appendOpaque(redirdest, 0, 0, 0);
 
   
-  TRACE(REQ, " XrdHttpReq::Redir Redirecting to " << redirdest.c_str());
+  TRACE(REQ, " XrdHttpReq::Redir Redirecting to " << obfuscateAuth(redirdest.c_str()).c_str());
 
   if (request != rtGET)
     prot->SendSimpleResp(307, NULL, (char *) redirdest.c_str(), 0, 0, keepalive);

--- a/src/XrdOuc/XrdOucUtils.cc
+++ b/src/XrdOuc/XrdOucUtils.cc
@@ -123,7 +123,7 @@ static const std::string OBFUSCATION_STR = "REDACTED";
 static const std::vector<std::regex> authObfuscationRegexes = {
   //authz=xxx&... We deal with cases like "(message: kXR_stat (path: /tmp/xrootd/public/foo?pelican.timeout=3s&authz=foo1234, flags: none)" where we do not want to obfuscate
   // ", flags: none)" + we deal with cases where the 'authz=Bearer token' when an admin could set 'http.header2cgi Authorization authz' in the server config
-  std::regex(R"((authz=)(Bearer\s)?([^ &\",<>#%{}|\^~\[\]`]*))",std::regex_constants::optimize),
+  std::regex(R"((authz=)(Bearer(\s|%20))?([^ &\",<>#%{}|\^~\[\]`]*))",std::regex_constants::optimize),
   // HTTP Authorization, TransferHeaderAuthorization headers that with the key that can be prefixed with spaces and value prefixed by spaces
   std::regex(R"((\s*\w*Authorization\s*:\s*)[^$]*)", std::regex_constants::icase | std::regex_constants::optimize)
 };

--- a/src/XrdPosix/XrdPosixFile.cc
+++ b/src/XrdPosix/XrdPosixFile.cc
@@ -38,6 +38,7 @@
 #include <sys/stat.h>
 
 #include "XrdOuc/XrdOucName2Name.hh"
+#include "XrdOuc/XrdOucPrivateUtils.hh"
 #include "XrdPosix/XrdPosixCallBack.hh"
 #include "XrdPosix/XrdPosixConfig.hh"
 #include "XrdPosix/XrdPosixFile.hh"
@@ -236,9 +237,9 @@ do{if (doWait)
                 }
              if (Say)
                 {snprintf(buff, sizeof(buff), "%s timeout closing", eTxt);
-                 Say->Emsg("DDestroy", buff, fCurr->Origin());
+                 Say->Emsg("DDestroy", buff, obfuscateAuth(fCurr->Origin()).c_str());
                 } else {
-                 DMSG("DDestroy", eTxt <<" timeout closing " <<fCurr->Origin()
+                 DMSG("DDestroy", eTxt <<" timeout closing " << obfuscateAuth(fCurr->Origin())
                         <<' ' <<ddNumLost <<" objects lost");
                 }
              fCurr->nextFile = ddLost;
@@ -295,9 +296,10 @@ void XrdPosixFile::DelayedDestroy(XrdPosixFile *fp)
    fp->numTries = 0;
    ddMutex.UnLock();
 
-   DEBUG("DLY destroy "<<(doPost ? "post " : "has ")<<ddCount
-                       <<" objects; added "<<fp->Origin());
-
+   if(DEBUGON) {
+     DEBUG("DLY destroy " << (doPost ? "post " : "has ") << ddCount
+                          << " objects; added " << obfuscateAuth(fp->Origin()));
+   }
    if (doPost) ddSem.Post();
 }
 

--- a/src/XrdPosix/XrdPosixPrepIO.cc
+++ b/src/XrdPosix/XrdPosixPrepIO.cc
@@ -28,6 +28,7 @@
 /* specific prior written permission of the institution or contributor.       */
 /******************************************************************************/
 
+#include "XrdOuc/XrdOucPrivateUtils.hh"
 #include "XrdPosix/XrdPosixObjGuard.hh"
 #include "XrdPosix/XrdPosixPrepIO.hh"
 #include "XrdPosix/XrdPosixStats.hh"
@@ -52,7 +53,9 @@ void XrdPosixPrepIO::Disable()
    EPNAME("PrepIODisable");
    XrdPosixObjGuard objGuard(fileP);
 
-   DEBUG("Disabling deferred open "<<fileP->Origin());
+   if(DEBUGON) {
+     DEBUG("Disabling deferred open " << obfuscateAuth(fileP->Origin()));
+   }
 
    openRC = -ESHUTDOWN;
 }
@@ -102,7 +105,7 @@ bool XrdPosixPrepIO::Init(XrdOucCacheIOCB *iocbP)
       else {openRC = XrdPosixMap::Result(Status, fileP->ecMsg, false);
             if (DEBUGON && errno != ENOENT && errno != ELOOP)
                {std::string eTxt = Status.ToString();
-                DEBUG(eTxt<<" deferred open "<<fileP->Origin());
+                DEBUG(eTxt<<" deferred open "<< obfuscateAuth(fileP->Origin()));
                }
             XrdPosixGlobals::Stats.Count(XrdPosixGlobals::Stats.X.OpenErrs);
             return false;

--- a/src/XrdPosix/XrdPosixXrootd.cc
+++ b/src/XrdPosix/XrdPosixXrootd.cc
@@ -55,6 +55,7 @@
 #include "XrdOuc/XrdOucEnv.hh"
 #include "XrdOuc/XrdOucName2Name.hh"
 #include "XrdOuc/XrdOucPsx.hh"
+#include "XrdOuc/XrdOucPrivateUtils.hh"
 
 #include "XrdPosix/XrdPosixAdmin.hh"
 #include "XrdPosix/XrdPosixCallBack.hh"
@@ -287,7 +288,7 @@ int XrdPosixXrootd::Close(int fildes)
       {if ((ret = fP->Close(Status))) {delete fP; fP = 0;}
           else if (DEBUGON)
                   {std::string eTxt = Status.ToString();
-                   DEBUG(eTxt <<" closing " <<fP->Origin());
+                   DEBUG(eTxt <<" closing " << obfuscateAuth(fP->Origin()));
                   }
       } else {
        ret = true;
@@ -634,7 +635,7 @@ int XrdPosixXrootd::Open(const char *path, int oflags, mode_t mode,
       {XrdPosixGlobals::Stats.Count(XrdPosixGlobals::Stats.X.OpenErrs);
        int rc = XrdPosixMap::Result(Status,XrdPosixGlobals::ecMsg,false);
        if (DEBUGON && rc != -ENOENT && rc != -ELOOP)
-          {DEBUG(XrdPosixGlobals::ecMsg.Msg() <<" open " <<fp->Origin());}
+          {DEBUG(XrdPosixGlobals::ecMsg.Msg() <<" open " << obfuscateAuth(fp->Origin()));}
        delete fp;
        errno = -rc;  // Saved errno across the delete
        return -1;

--- a/src/XrdPss/XrdPss.cc
+++ b/src/XrdPss/XrdPss.cc
@@ -64,6 +64,7 @@
 #include "XrdOuc/XrdOucEnv.hh"
 #include "XrdOuc/XrdOucExport.hh"
 #include "XrdOuc/XrdOucPgrwUtils.hh"
+#include "XrdOuc/XrdOucPrivateUtils.hh"
 #include "XrdSec/XrdSecEntity.hh"
 #include "XrdSecsss/XrdSecsssID.hh"
 #include "XrdSfs/XrdSfsInterface.hh"
@@ -363,7 +364,10 @@ int XrdPssSys::Mkdir(const char *path, mode_t mode, int mkpath, XrdOucEnv *eP)
 
 // Some tracing
 //
-   DEBUG(uInfo.Tident(),"url="<<pbuff);
+  if(DEBUGON) {
+    auto urlObf = obfuscateAuth(pbuff);
+    DEBUG(uInfo.Tident(),"url="<<urlObf);
+  }
 
 // Simply return the proxied result here
 //
@@ -407,8 +411,10 @@ int XrdPssSys::Remdir(const char *path, int Opts, XrdOucEnv *eP)
 
 // Do some tracing
 //
-   DEBUG(uInfo.Tident(),"url="<<pbuff);
-
+  if(DEBUGON) {
+    auto urlObf = obfuscateAuth(pbuff);
+    DEBUG(uInfo.Tident(),"url="<<urlObf);
+  }
 // Issue unlink and return result
 //
    return (XrdPosixXrootd::Rmdir(pbuff) ? -errno : XrdOssOK);
@@ -451,7 +457,12 @@ int XrdPssSys::Rename(const char *oldname, const char *newname,
 
 // Do some tracing
 //
-   DEBUG(uInfoOld.Tident(),"old url="<<oldName <<" new url=" <<newName);
+  if(DEBUGON) {
+    auto oldNameObf = obfuscateAuth(oldName);
+    auto newNameObf = obfuscateAuth(newName);
+    DEBUG(uInfoOld.Tident(),"old url="<<oldNameObf <<" new url=" <<newNameObf);
+  }
+
 
 // Execute the rename and return result
 //
@@ -504,7 +515,10 @@ int XrdPssSys::Stat(const char *path, struct stat *buff, int Opts, XrdOucEnv *eP
 
 // Do some tracing
 //
-   DEBUG(uInfo.Tident(),"url="<<pbuff);
+  if(DEBUGON) {
+    auto urlObf = obfuscateAuth(pbuff);
+    DEBUG(uInfo.Tident(),"url="<<urlObf);
+  }
 
 // Return proxied stat
 //
@@ -560,7 +574,10 @@ int XrdPssSys::Truncate(const char *path, unsigned long long flen,
 
 // Do some tracing
 //
-   DEBUG(uInfo.Tident(),"url="<<pbuff);
+  if(DEBUGON) {
+    auto urlObf = obfuscateAuth(pbuff);
+    DEBUG(uInfo.Tident(),"url="<<urlObf);
+  }
 
 // Return proxied truncate. We only do this on a single machine because the
 // redirector will forbid the trunc() if multiple copies exist.
@@ -605,7 +622,10 @@ int XrdPssSys::Unlink(const char *path, int Opts, XrdOucEnv *envP)
 
 // Do some tracing
 //
-   DEBUG(uInfo.Tident(),"url="<<pbuff);
+  if(DEBUGON) {
+    auto urlObf = obfuscateAuth(pbuff);
+    DEBUG(uInfo.Tident(),"url="<<urlObf);
+  }
 
 // Unlink the file and return result.
 //
@@ -653,7 +673,10 @@ int XrdPssDir::Opendir(const char *dir_path, XrdOucEnv &Env)
 
 // Do some tracing
 //
-   DEBUG(uInfo.Tident(),"url="<<pbuff);
+  if(DEBUGON) {
+    auto urlObf = obfuscateAuth(pbuff);
+    DEBUG(uInfo.Tident(),"url="<<urlObf);
+  }
 
 // Open the directory
 //
@@ -845,7 +868,10 @@ int XrdPssFile::Open(const char *path, int Oflag, mode_t Mode, XrdOucEnv &Env)
 
 // Do some tracing
 //
-   DEBUG(uInfo.Tident(),"url="<<pbuff);
+  if(DEBUGON) {
+    auto urlObf = obfuscateAuth(pbuff);
+    DEBUG(uInfo.Tident(),"url="<<urlObf);
+  }
 
 // Try to open and if we failed, return an error
 //

--- a/src/XrdPss/XrdPssCks.cc
+++ b/src/XrdPss/XrdPssCks.cc
@@ -37,6 +37,7 @@
 #include "XrdVersion.hh"
 
 #include "XrdOuc/XrdOucTokenizer.hh"
+#include "XrdOuc/XrdOucPrivateUtils.hh"
 #include "XrdPosix/XrdPosixXrootd.hh"
 #include "XrdPss/XrdPssUrlInfo.hh"
 
@@ -129,7 +130,11 @@ int XrdPssCks::Get(const char *Pfn, XrdCksData &Cks)
 
 // Do some debugging
 //
-    DEBUG(uInfo.Tident(),"url="<<pBuff);
+
+    if(DEBUGON) {
+      auto urlObf = obfuscateAuth(pBuff);
+      DEBUG(uInfo.Tident(),"url="<<urlObf);
+    }
 
 // First step is to getthe checksum value
 //

--- a/src/XrdPss/XrdPssTrace.hh
+++ b/src/XrdPss/XrdPssTrace.hh
@@ -39,8 +39,12 @@
 
 #define QTRACE(act) SysTrace.What & TRACEPSS_ ## act
 
+#define TRACING(x) (SysTrace.What & x)
+
+#define DEBUGON TRACING(TRACEPSS_Debug)
+
 #define DEBUG(tid,y)\
-    if (SysTrace.What & TRACEPSS_Debug) SYSTRACE(SysTrace., tid, epname, 0, y)
+    if (DEBUGON) SYSTRACE(SysTrace., tid, epname, 0, y)
 
 #define EPNAME(x) static const char *epname = x;
 

--- a/tests/XrdOucTests/XrdOucUtilsTests.cc
+++ b/tests/XrdOucTests/XrdOucUtilsTests.cc
@@ -41,6 +41,7 @@ TEST(XrdOucUtilsTests, obfuscateAuth) {
   ASSERT_EQ(std::string("Appended header field to opaque info: 'authz=") + OBFUSCATION_STR, obfuscateAuth("Appended header field to opaque info: 'authz=Bearer abcdef'"));
   ASSERT_EQ(std::string("Appended header fields to opaque info: 'authz=") + OBFUSCATION_STR + std::string("&scitag.flow=65'"), obfuscateAuth("Appended header fields to opaque info: 'authz=Bearer token&scitag.flow=65'"));
   ASSERT_EQ(std::string("Processing source entry: /etc/passwd, type local file, target file: root://localhost:1094//tmp/passwd?authz=") + OBFUSCATION_STR, obfuscateAuth("Processing source entry: /etc/passwd, type local file, target file: root://localhost:1094//tmp/passwd?authz=testabcd"));
+  ASSERT_EQ(std::string("240919 08:11:07 20995 unknown.3:33@[::1] Pss_Stat: url=pelican://p0@F4HP7QL65F.local:61631//first/namespace/token_gen/test1171569942?&authz=") + OBFUSCATION_STR + std::string("&pelican.timeout=9.5s"),obfuscateAuth("240919 08:11:07 20995 unknown.3:33@[::1] Pss_Stat: url=pelican://p0@F4HP7QL65F.local:61631//first/namespace/token_gen/test1171569942?&authz=Bearer%20tokentoberedacted&pelican.timeout=9.5s"));
 }
 
 TEST(XrdOucUtilsTests, caseInsensitiveFind) {

